contracts/modules/TransferManager/LockUpTransferManager.sol

```diff
@@ -0,0 +1,651 @@
+     */
+    function _checkIfValidTransfer(address _userAddress, uint256 _amount) internal view returns (Result) {
+        uint256 totalRemainingLockedAmount = getLockedTokenToUser(_userAddress);
+        // Present balance of the user
+        uint256 currentBalance = ISecurityToken(securityToken).balanceOf(_userAddress);
+        if ((currentBalance.sub(_amount)) >= totalRemainingLockedAmount) {
+            return Result.NA;
+        }
+        return Result.INVALID;
+    }
+
+    /**
+     * @notice Provide the unlock amount for the given lockup for a particular user
+     */
+    function _getUnlockedAmountForLockup(bytes32 _lockupName) internal view returns (uint256) {
+        /*solium-disable-next-line security/no-block-members*/
+        if (lockups[_lockupName].startTime > now) {
+            return 0;
+        } else if (lockups[_lockupName].startTime.add(lockups[_lockupName].lockUpPeriodSeconds) <= now) {
+            return lockups[_lockupName].lockupAmount;
+        } else {
+            // Calculate the no. of periods for a lockup 
+            uint256 noOfPeriods = (lockups[_lockupName].lockUpPeriodSeconds).div(lockups[_lockupName].releaseFrequencySeconds);
+            // Calculate the transaction time lies in which period
+            /*solium-disable-next-line security/no-block-members*/
+            uint256 elapsedPeriod = (now.sub(lockups[_lockupName].startTime)).div(lockups[_lockupName].releaseFrequencySeconds);
+            // Calculate the allowed unlocked amount per period
+            uint256 amountPerPeriod = (lockups[_lockupName].lockupAmount).div(noOfPeriods);
+            // Find out the unlocked amount for a given lockup
+            uint256 unLockedAmount = elapsedPeriod.mul(amountPerPeriod);
+            return unLockedAmount;
+        } 
+    }
+
+    function _removeLockupType(bytes32 _lockupName) internal {
+        require(lockups[_lockupName].startTime != 0, "Lockup type doesn’t exist");
+        require(lockupToUsers[_lockupName].length == 0, "Users are associated with the lockup");
+        // delete lockup type 
+        delete(lockups[_lockupName]);
+        uint256 i = 0;
+        for (i = 0; i < lockupArray.length; i++) {
+            if (lockupArray[i] == _lockupName) {
+                break;
+            }
+        }
+        if (i != lockupArray.length -1) {
+            lockupArray[i] = lockupArray[lockupArray.length -1];
+        }
+        lockupArray.length--;
+        emit RemoveLockUpType(_lockupName);
+    }
+
+    function _modifyLockUpType(
+        uint256 _lockupAmount,
+        uint256 _startTime,
+        uint256 _lockUpPeriodSeconds,
+        uint256 _releaseFrequencySeconds,
+        bytes32 _lockupName
+    )
+        internal
+    {
+        /*solium-disable-next-line security/no-block-members*/
+        require(lockups[_lockupName].startTime > now, "Not allowed to modify");
+        require(_startTime >= now, "Past time is not allowed");
```

Comment:
Yes, it is fine to give upper bound time to mine. otherwise, Issuer will able to create 6 months back lockup or more.

Reply:
What is the problem with issuer creating a lockup that started 6 months before?

Reply:
It doesn't make sense. If I am the user, I was able to transfer the tokens freely from last 6 months but on the Issuer portal, it will say the user was restricted from last 6 months while he already sells more than the lockup amount freely.

Reply:
If an issuer is adding a lockup in past, he will be well aware of this. 

Anyway, how about we treat `_startTime = 0` as `_startTime = now` so that if an issuer wants to immediately start a lockup, he can do so. It will solve the UX problem I mentioned above.

Reply:
we could do that